### ISSUE TITLE
Issue using AutoCommitMarks and MarkCommitRecords

### ISSUE THREAD

**Issue by cnmota** [twmb/franz-go#125]
Hi,

First let me thank you for an excellent package, now regarding the issue and I might be using the package incorrectly but when implementing a consumer using AutoCommitMarks and MarkCommitRecords and consuming a topic from the start the consumer group offsets are not being moved forward.

From what I understood of the code the issue happens because AutoCommitMarks used offset.head as the value of the commit and at the start of the topic dirty has a value but head is at 0.

I managed to fix this by change getUncommittedLocked from

```
func (g *groupConsumer) getUncommittedLocked(head, dirty bool) map[string]map[int32]EpochOffset {
	if g.uncommitted == nil {
		return nil
	}

	var uncommitted map[string]map[int32]EpochOffset
	for topic, partitions := range g.uncommitted {
		var topicUncommitted map[int32]EpochOffset
		for partition, uncommit := range partitions {
			if head && uncommit.dirty == uncommit.committed {
				continue
			}
			if topicUncommitted == nil {
				if uncommitted == nil {
					uncommitted = make(map[string]map[int32]EpochOffset, len(g.uncommitted))
				}
				topicUncommitted = uncommitted[topic]
				if topicUncommitted == nil {
					topicUncommitted = make(map[int32]EpochOffset, len(partitions))
					uncommitted[topic] = topicUncommitted
				}
			}
			fmt.Println(partition, "dirty:", uncommit.dirty, "head:", uncommit.head)
			if head {
				if dirty {
					topicUncommitted[partition] = uncommit.dirty
				} else {
					topicUncommitted[partition] = uncommit.head
				}
			} else {
				topicUncommitted[partition] = uncommit.committed
			}
		}
	}

	return uncommitted
}
```

to

```
func (g *groupConsumer) getUncommittedLocked(head, dirty bool) map[string]map[int32]EpochOffset {
	if g.uncommitted == nil {
		return nil
	}

	var uncommitted map[string]map[int32]EpochOffset
	for topic, partitions := range g.uncommitted {
		var topicUncommitted map[int32]EpochOffset
		for partition, uncommit := range partitions {
			if head && uncommit.dirty == uncommit.committed {
				continue
			}
			if topicUncommitted == nil {
				if uncommitted == nil {
					uncommitted = make(map[string]map[int32]EpochOffset, len(g.uncommitted))
				}
				topicUncommitted = uncommitted[topic]
				if topicUncommitted == nil {
					topicUncommitted = make(map[int32]EpochOffset, len(partitions))
					uncommitted[topic] = topicUncommitted
				}
			}
			fmt.Println(partition, "dirty:", uncommit.dirty, "head:", uncommit.head)
			if head {
				if dirty || (uncommit.head == EpochOffset{}) {
					topicUncommitted[partition] = uncommit.dirty
				} else {
					topicUncommitted[partition] = uncommit.head
				}
			} else {
				topicUncommitted[partition] = uncommit.committed
			}
		}
	}

	return uncommitted
}
```

so essentially adding `(uncommit.head == EpochOffset{})` this because commit from AutoCommitMarks appears to always use only head for the initial commit. Like I said there is probably a better place to make this fix or I might be doing something wrong.

some additional info, starting the consumer with this settings

```
		kgo.SeedBrokers(c.Config.Brokers...),
		kgo.ConsumerGroup(group),
		kgo.AutoCommitMarks(),
		kgo.WithLogger(logWrapper{writer: c.logger}),
		kgo.OnPartitionsAssigned(d.assignPartitions),
		kgo.OnPartitionsRevoked(d.removePartitions),
		kgo.OnPartitionsLost(d.removePartitions),
```

**Comment by twmb** [comment]
I'm not sure I'm understanding here, because `MarkCommitRecords` sets the `head` on uncommitted offsets -- when using marks, the client never wants to use dirty offsets, and when autocommitting happens, the head will be used (which, through marking, grabs any `uncommit` currently marked head).

What behavior are you testing to see offsets not advancing?

**Comment by cnmota** [comment]
Sorry if I did not explain myself correctly, what I am observing is that offsets are not advancing, but this only happens when I start consuming on a newly auto created consumer group some additional feedback:

```
kafka-consumer-groups --bootstrap-server=$KPRD --group=delme-range-transformer --delete
Deletion of requested consumer groups ('delme-range-transformer') was successful.
```

Now if I start my consumer for that group with MarkCommitRecords and AutoCommitMarks we get the following debug logs:

```
{"timestamp":"2022-01-28T11:45:37.470658Z","msg":"wrote Fetch v10 [broker: 2, bytes_written: 104, write_wait: 21.203µs, time_to_write: 14.221µs, err: <nil>]"}
{"timestamp":"2022-01-28T11:45:37.495965Z","msg":"read Fetch v10 [broker: 1, bytes_read: 131158, read_wait: 85.366µs, time_to_read: 44.198271ms, err: <nil>]"}
{"timestamp":"2022-01-28T11:45:37.496653Z","msg":"updated uncommitted [group: delme-range-transformer, to: golem-raw-update-log[1{0=>121847=>122134}]]"}
{"timestamp":"2022-01-28T11:45:37.496856Z","msg":"wrote Fetch v10 [broker: 1, bytes_written: 104, write_wait: 77.231µs, time_to_write: 27.34µs, err: <nil>]"}
{"timestamp":"2022-01-28T11:45:37.503037Z","msg":"read Fetch v10 [broker: 3, bytes_read: 131158, read_wait: 65.38µs, time_to_read: 86.093573ms, err: <nil>]"}
{"timestamp":"2022-01-28T11:45:37.503454Z","msg":"updated uncommitted [group: delme-range-transformer, to: golem-raw-update-log[3{0=>265=>550}]]"}
{"timestamp":"2022-01-28T11:45:37.50355Z","msg":"wrote Fetch v10 [broker: 3, bytes_written: 104, write_wait: 29.748µs, time_to_write: 21.583µs, err: <nil>]"}
{"timestamp":"2022-01-28T11:45:37.511574Z","msg":"read Fetch v10 [broker: 2, bytes_read: 131158, read_wait: 40.213µs, time_to_read: 40.866234ms, err: <nil>]"}
{"timestamp":"2022-01-28T11:45:37.511936Z","msg":"updated uncommitted [group: delme-range-transformer, to: golem-raw-update-log[2{0=>2919=>3181}]]"}
{"timestamp":"2022-01-28T11:45:37.512021Z","msg":"wrote Fetch v10 [broker: 2, bytes_written: 104, write_wait: 30.28µs, time_to_write: 16.514µs, err: <nil>]"}
{"timestamp":"2022-01-28T11:45:37.537064Z","msg":"read Fetch v10 [broker: 1, bytes_read: 131158, read_wait: 97.468µs, time_to_read: 40.118467ms, err: <nil>]"}
{"timestamp":"2022-01-28T11:45:37.55176Z","msg":"read Fetch v10 [broker: 2, bytes_read: 131158, read_wait: 35.635µs, time_to_read: 39.691381ms, err: <nil>]"}
{"timestamp":"2022-01-28T11:45:37.57757Z","msg":"read Fetch v10 [broker: 3, bytes_read: 131158, read_wait: 50.985µs, time_to_read: 73.949995ms, err: <nil>]"}
{"timestamp":"2022-01-28T11:45:38.142698Z","msg":"autocommitting [group: delme-range-transformer]"}
{"timestamp":"2022-01-28T11:45:38.142839Z","msg":"issuing commit [group: delme-range-transformer, uncommitted: map[golem-raw-update-log:map[0:{0 0} 1:{0 0} 2:{0 0} 3:{0 0}]]]"}
```

From what I can understand MarkCommitRecords sets the dirty correctly but head is set at **0** and when `getUncommittedLocked` is then called it's called with dirty as false hence entering here:

```
			if head {
				if dirty {
					topicUncommitted[partition] = uncommit.dirty
				} else {
					topicUncommitted[partition] = uncommit.head
				}
			} else {
				topicUncommitted[partition] = uncommit.committed
			}
```

Thus returning the value of head which is 0 so AutoCommit will receive only zeros as offsets and will commit that and never ended up moving forward head is always 0 forever.

Like I said before if the initial offset of the consumer group on partitions is already set with something other than 0 it will commit normally and the topic will move forward.

**Comment by twmb** [comment]
Is it possible to have a small example of your code?

In the client, `MarkCommitRecords` updates the `head` itself if the current `head` is less than what's being committed: https://github.com/twmb/franz-go/blob/0a23cca3f4ee9d69b6cb2e70aab1c8e012b901ad/pkg/kgo/consumer_group.go#L1939-L1941 (same logic for dirty, but that is less important here). I remember testing this when I implemented it (initial version had a problem), but I can't test it at the moment -- do not have access to my laptop that runs a small cluster. I'll be able to check this again next week, but I wonder if we can suss the issue out before that.

**Comment by cnmota** [comment]
Hi, of course I can, will add it tomorrow I don’t have access at the moment to my laptop but tomorrow I can add a sample of the code, and will also add some prints of what the snippet you posted is working with.

thank you !

**Comment by cnmota** [comment]
Hi here is the whole package package that we are using as a driver implementation for Franz we where using sarama mostly and are slowly trying to migrate to franz.

```package consumers

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/twmb/franz-go/pkg/kgo"
	"go.worten.net/digital/packages/pubsub"
	"go.worten.net/digital/packages/pubsub/drivers/josef/internal/adapters"
	"go.worten.net/digital/packages/pubsub/drivers/josef/internal/configs"
	"go.worten.net/digital/packages/pubsub/drivers/josef/internal/translators"
)

type logWrapper struct {
	writer adapters.Logger
}

func (l logWrapper) Level() kgo.LogLevel {
	return kgo.LogLevelError
}

func (l logWrapper) Log(level kgo.LogLevel, msg string, keyvals ...interface{}) {
	if len(keyvals) != 0 && len(keyvals)%2 == 0 {
		format := strings.Repeat("%v: %v, ", len(keyvals)/2)
		format = format[:len(format)-2] // trim trailing comma and space
		msg += fmt.Sprintf(" ["+format+"]", keyvals...)
	}

	l.writer.Log(msg)
}

// FranzGroupConsumer is the implementation of a group consumer.
type FranzGroupConsumer struct {
	Config configs.Consumer
	logger adapters.Logger
}

func (c FranzGroupConsumer) Connect(ctx context.Context, topic, group string) error {
	if err := c.Config.Sanitize(); err != nil {
		return err
	}

	c.logger = c.Config.Logger

	d := newDispatcher(c.Config, c.logger, consumerOptions{commitOnFlush: true})

	opts := []kgo.Opt{
		kgo.SeedBrokers(c.Config.Brokers...),
		kgo.ConsumerGroup(group),
		kgo.AutoCommitMarks(),
		kgo.WithLogger(logWrapper{writer: c.logger}),
		kgo.OnPartitionsAssigned(d.assignPartitions),
		kgo.OnPartitionsRevoked(d.removePartitions),
		kgo.OnPartitionsLost(d.removePartitions),
	}

	if c.Config.InitialOffset != "" {
		offset, err := translators.FranzOffsetFromString(c.Config.InitialOffset)
		if err != nil {
			return fmt.Errorf("failed to infer initial offset on group consumer: %w", err)
		}

		opts = append(opts, kgo.ConsumeResetOffset(offset))
	}

	opts = append(opts, c.Config.ToFranzConfig(topic)...)

	cl, err := kgo.NewClient(opts...)
	if err != nil {
		return err
	}
	defer cl.Close()

	return d.dispatch(ctx, cl)
}

// FranzGroupConsumer is the implementation of a group consumer.
type FranzPartitionConsumer struct {
	Config configs.Consumer
	logger adapters.Logger
}

func (c FranzPartitionConsumer) Connect(ctx context.Context, topic string) error {
	if err := c.Config.Sanitize(); err != nil {
		return err
	}

	c.logger = c.Config.Logger

	var partitions = make([]int32, 0, len(c.Config.Offsets))
	var offsets = make(map[int32]kgo.Offset, len(c.Config.Offsets))

	for partitionID, offset := range c.Config.Offsets {
		partitions = append(partitions, partitionID)
		offsets[partitionID] = translators.FranzOffsetFromOffset(offset)
	}

	opts := []kgo.Opt{
		kgo.SeedBrokers(c.Config.Brokers...),
		kgo.WithLogger(logWrapper{writer: c.logger}),
		kgo.ConsumePartitions(map[string]map[int32]kgo.Offset{topic: offsets}),
	}

	opts = append(opts, c.Config.ToFranzConfig(topic)...)

	cl, err := kgo.NewClient(opts...)
	if err != nil {
		return err
	}
	defer cl.Close()

	d := newDispatcher(c.Config, c.logger, consumerOptions{})

	d.assignPartitions(ctx, cl, map[string][]int32{topic: partitions})

	return d.dispatch(ctx, cl)
}

type dispatcher struct {
	mu              sync.Mutex
	logger          adapters.Logger
	config          configs.Consumer
	consumerOptions consumerOptions
	consumers       map[string]map[int32]partitionConsumer
}

func newDispatcher(cfg configs.Consumer, l adapters.Logger, o consumerOptions) *dispatcher {
	d := &dispatcher{
		config:          cfg,
		logger:          l,
		consumerOptions: o,
		consumers:       make(map[string]map[int32]partitionConsumer),
	}

	return d
}

func (d *dispatcher) assignPartitions(ctx context.Context, cl *kgo.Client, assigned map[string][]int32) {
	d.mu.Lock()
	defer d.mu.Unlock()

	for topic, partitions := range assigned {
		if d.consumers[topic] == nil {
			d.consumers[topic] = make(map[int32]partitionConsumer)
		}

		for _, partition := range partitions {
			d.logger.Log("assigning [topic:%s][partition:%d]", topic, partition)

			pc := newPartitionConsumer(cl, d.logger, d.config, d.consumerOptions)
			d.consumers[topic][partition] = pc

			go pc.consume(ctx, topic, partition)
		}
	}
}

func (d *dispatcher) removePartitions(ctx context.Context, cl *kgo.Client, lost map[string][]int32) {
	d.mu.Lock()
	defer d.mu.Unlock()

	for topic, partitions := range lost {
		ptopics := d.consumers[topic]

		for _, partition := range partitions {
			d.logger.Log("unassigning [topic:%s][partition:%d]", topic, partition)

			pc := ptopics[partition]
			delete(ptopics, partition)

			if len(ptopics) == 0 {
				delete(d.consumers, topic)
			}

			close(pc.quit)
		}
	}
}

func (d *dispatcher) dispatch(ctx context.Context, cl *kgo.Client) error {
	for {
		fetches := cl.PollFetches(ctx)
		if fetches.IsClientClosed() {
			return nil
		}

		for _, err := range fetches.Errors() {
			return fmt.Errorf("error reading topic %s, partition %d: %w", err.Topic, err.Partition, err.Err)
		}

		fetches.EachTopic(func(t kgo.FetchTopic) {
			d.mu.Lock()
			tconsumers := d.consumers[t.Topic]
			d.mu.Unlock()

			if tconsumers == nil {
				return
			}

			t.EachPartition(func(p kgo.FetchPartition) {
				pc, ok := tconsumers[p.Partition]
				if !ok {
					return
				}

				select {
				case pc.recs <- p.Records:
				case <-pc.quit:
				}
			})
		})

		select {
		case <-ctx.Done():
			return nil
		default:
			continue
		}

	}
}

type consumerOptions struct {
	commitOnFlush bool
}

type partitionConsumer struct {
	client            *kgo.Client
	options           consumerOptions
	logger            adapters.Logger
	onHeartbeat       func(ctx context.Context, topic string, partition int32, lastOffset, lag int64) error
	handleMessages    func(ctx context.Context, messages ...pubsub.Envelope) error
	backoff           time.Duration
	bulkSize          int
	bulkGatherTimeout time.Duration

	quit chan struct{}
	recs chan []*kgo.Record
}

func newPartitionConsumer(cli *kgo.Client, l adapters.Logger, cfg configs.Consumer, o consumerOptions) partitionConsumer {
	return partitionConsumer{
		client:            cli,
		logger:            l,
		options:           o,
		onHeartbeat:       cfg.OnHeartbeat,
		handleMessages:    cfg.Handler,
		bulkSize:          cfg.BulkSize,
		bulkGatherTimeout: cfg.BulkGatherTimeout,
		backoff:           cfg.Backoff,
		quit:              make(chan struct{}),
		recs:              make(chan []*kgo.Record, 10),
	}
}

func (pc *partitionConsumer) consume(ctx context.Context, topic string, partition int32) {
	var (
		ticker    = time.NewTicker(pc.bulkGatherTimeout)
		heartbeat = time.NewTicker(5 * time.Second)
		batch     = NewBatch(topic, partition, pc.bulkSize)
	)
	defer ticker.Stop()
	defer heartbeat.Stop()

	for {
		select {
		case <-ctx.Done():
			pc.logger.Log("terminating partition consumer, context terminated [topic:%s][partition:%d]", topic, partition)
			return
		case <-pc.quit:
			pc.logger.Log("terminating partition consumer, partition re-assigned [topic:%s][partition:%d]", topic, partition)
			return
		case <-heartbeat.C:
			if pc.onHeartbeat == nil {
				continue
			}
			if err := pc.onHeartbeat(ctx, batch.Topic, batch.Partition, batch.LastOffset, 0); err != nil {
				pc.logger.Log("failed to issue heartbeat [topic:%s][partition:%d]: %w", topic, partition, err)
			}
		case recs := <-pc.recs:
			for _, rec := range recs {
				if batch.AddMessage(translators.FromFranzRecord(rec), rec.Offset) < pc.bulkSize {
					continue
				}

				if err := pc.flush(ctx, batch); err != nil {
					pc.logger.Log("failed to commit offsets on [topic:%s][partition:%d]: %w", topic, partition, err)
				}
			}
		case <-ticker.C:
			if err := pc.flush(ctx, batch); err != nil {
				pc.logger.Log("failed to commit offsets on [topic:%s][partition:%d]: %w", topic, partition, err)
			}
		}
	}
}

func (pc *partitionConsumer) flush(ctx context.Context, batch *Batch) error {
	var messages = batch.Messages

	if len(messages) == 0 {
		return nil
	}

	for err := pc.handleMessages(ctx, messages...); err != nil; err = pc.handleMessages(ctx, messages...) {
		select {
		case <-ctx.Done():
			return nil
		case <-time.After(pc.backoff):
			continue
		}
	}

	if pc.options.commitOnFlush {
		pc.client.MarkCommitRecords(&kgo.Record{
			Topic: batch.Topic, Partition: batch.Partition, Offset: batch.LastOffset, LeaderEpoch: -1,
		})
	}

	batch.Reset()

	return nil
}```

For the issue I mentioned essentially we are using 

```func (c FranzGroupConsumer) Connect(ctx context.Context, topic, group string) error {```

If I replace MarkCommitRecords and AutoCommitMarks with DisableAutoCommits and CommitRecords the problem does not happen.

Some additional notes:

* We are using version 1.2.6. of Franz
* Don't know if LeadeEpochOffset:: -1 can be causing the issue.

On the following message will add print outs of:

```
 if next.head.less(set) { 
 	next.head = set 
 } 
```

When the problem happens

**Comment by cnmota** [comment]
So first we completely delete the consumer group:

```
backend/golem - [hotfixes/range-transformer-tunning●] » kafka-consumer-groups --bootstrap-server=$KPRD --group=delme-range-transformer --describe

Error: Consumer group 'delme-range-transformer' does not exist.
backend/golem - [hotfixes/range-transformer-tunning●] » 
```

I changed consumer_group.go to print out and wait a bit just to understand what is happening on the snippet you mentioned:

```
		next := curPartitions[r.Partition]
		fmt.Println("next head is:", next.head, "set is:", set)
		fmt.Println("next dirty is:", next.dirty, "set is:", set)
		if next.head.less(set) {
			fmt.Println("setting head since head less tan set")
			next.head = set
		}
		if next.dirty.less(set) { // for sanity, but this should not happen
			fmt.Println("setting dirty since dirty less tan set")
			next.dirty = set
		}
```

Now I start up the consumer and we get this output:

```
{"timestamp":"2022-01-29T17:04:11.161813Z","msg":"updated uncommitted [group: delme-range-transformer, to: golem-raw-update-log[2{0=>0=>280}]]"}
next head is: {0 0} set is: {-1 2}
next dirty is: {3 280} set is: {-1 2}
```
The -1 is LeaderEpoch which is explicitly set as -1 since we don't re-use the rs record, so according to the code

0 is never less than -1, so in theory if pass the correct LeaderEpoch it should work ... giving it a try

**Comment by cnmota** [comment]
Hi again I need to further test this but yes the problem seems to be `LeaderEpoch :-1` if I store the received LeaderEpoch from `kgo.Record` on our batch structure and pass it to MarkCommitRecords Group offsets are now working.

So the issue was passing LeaderEpoch as -1 that essentially borks up the `less` function, guess LeaderEpoch -1 is to be used with care and avoided.

Thank you for the pointers  ... want me to close the issue ?

**Comment by twmb** [comment]
Based on what you pointed out, the current mark code is structured to prevent rewinds. This is an artificial limitation and can be removed, which will allow you to use `-1` for the leader epoch (and will allow rewinding through MarkCommitRecords, which I originally sought to avoid). However, this will also require that you modify the leader epoch in OnPartitionsLost if you decide to manually commit there. I recommend keeping the leader epoch, as it allows consumers to detect data loss. Out of curiosity, why remove it?

Also, I see that your code extended my goroutine_per_partition example -- I may go back and attempt to simplify some of the example. I'm not sure what a wrapper would look like yet nor when I would have time, but yeah that's an area that could be improved.

**Comment by cnmota** [comment]
"However, this will also require that you modify the leader epoch in OnPartitionsLost if you decide to manually commit there. I recommend keeping the leader epoch, as it allows consumers to detect data loss. Out of curiosity, why remove it?"

Essentially because we built an abstraction layer for kafka drivers with a common message envelope so kgo.Record is converted to that type and since we started with sarama that only requires the offset our public message type did not support additional meta data for the message offset (LeaderEpoch for example). So I tried to find a quick solution of not having to change the Message struct to include the LeaderEpoch thus using the -1 directly on the call to `MarkCommitRecords`

But yesterday while trying out keeping LeaderEpoch I think I found a way of not being intrusive with the current Message type and keep LeaderEpoch for the record we pretend to commit. Since we use and additional Batch Struct that is private and can be used internally to keep track of the original kgo.Record metadata and then have it available at the MarkCommitRecords call.

I think with this last touch we can finally slowly deprecate usage of the sarama implementation.

And thanks for you examples they are very helpful.
